qiskit/algorithms/minimum_eigen_solvers/qaoa/qaoa_circuit.py

```diff
@@ -0,0 +1,209 @@
+        elif self._mixer_operator is None:
+            self._num_parameters = 2 * self._param_p
+            self._bounds = [(None, None)] * self._param_p + [(0, 2 * np.pi)] * self._param_p
+            # Mixer is just a sum of single qubit X's on each qubit. Evolving by this operator
+            # will simply produce rx's on each qubit.
+            mixer_terms = [(I ^ left) ^ X ^ (I ^ (self._num_qubits - left - 1))
+                           for left in range(self._num_qubits)]
+            self._mixer = sum(mixer_terms)
+
+    def _construct_circuit(self, parameters) -> QuantumCircuit:
+        """Construct a parametrized circuit."""
+        if not len(parameters) == self._num_parameters:
+            raise ValueError('Incorrect number of angles: expecting {}, but {} given.'.format(
+                self._num_parameters, len(parameters)
+            ))
+
+        # initialize circuit, possibly based on given register/initial state
+        if isinstance(self._initial_state, QuantumCircuit):
+            circuit_op = CircuitStateFn(self._initial_state)
+        else:
+            circuit_op = (H ^ self._num_qubits)
+
+        # iterate over layers
+        for idx in range(self._param_p):
+            # the first [:self._p] parameters are used for the cost operator,
+            # so we apply them here
+            circuit_op = (self._cost_operator * parameters[idx]).exp_i().compose(circuit_op)
+            if isinstance(self._mixer, OperatorBase):
+                mixer = cast(OperatorBase, self._mixer)
+                # we apply beta parameter in case of operator based mixer.
+                circuit_op = (mixer * parameters[idx + self._param_p]).exp_i().compose(circuit_op)
+            else:
+                # mixer as a quantum circuit that can be parameterized
+                mixer = cast(QuantumCircuit, self._mixer)
+                num_params = mixer.num_parameters
+                # the remaining [self._p:] parameters are used for the mixer,
+                # there may be multiple layers, so parameters are grouped by layers.
+                param_values = parameters[self._param_p + num_params * idx:
+                                          self._param_p + num_params * (idx + 1)]
+                param_dict = dict(zip(mixer.parameters, param_values))
+                mixer = mixer.assign_parameters(param_dict)
+                circuit_op = CircuitOp(mixer).compose(circuit_op)
+
+        evolution = EvolutionFactory.build(self._cost_operator)
+        circuit_op = evolution.convert(circuit_op)
+        return circuit_op.to_circuit()
+
+    @property
+    def cost_operator(self) -> OperatorBase:
+        """Returns an operator representing the cost of the optimization problem."""
+        return self._cost_operator
+
+    @cost_operator.setter
+    def cost_operator(self, cost_operator: OperatorBase) -> None:
+        """Sets cost operator."""
+        self._cost_operator = cost_operator
+        self._data = None
```

Comment:
Is there are reason not to call invalidate in base class which resets data to None. This way whatever else is needed by the superclasses to be invalidated if the circuit should be wiped gets done.

Reply:
Good point, was not aware of `_invalidate`.